File: dotnet/src/DocGen/Exceptions/RateLimitException.cs
namespace Dokmatiq.DocGen.Exceptions;

/// <summary>
/// Exception thrown on HTTP 429 – rate limit exceeded.
/// </summary>
public class RateLimitException : ApiException
{
    /// <summary>Seconds until the next request is allowed.</summary>
    public double? RetryAfter { get; }

    /// <summary>Total rate limit.</summary>
    public int? Limit { get; }

    /// <summary>Remaining requests in the current window.</summary>
    public int? Remaining { get; }

    public RateLimitException(string message, double? retryAfter, int? limit, int? remaining)
        : base(429, message)
    {
        RetryAfter = retryAfter;
        Limit = limit;
        Remaining = remaining;
    }
}

File: java/src/main/java/com/dokmatiq/docgen/exception/RateLimitException.java
package com.dokmatiq.docgen.exception;

/** 429 – Rate limit exceeded. */
public class RateLimitException extends ApiException {
    private final Double retryAfter;
    private final Integer limit;
    private final Integer remaining;

    public RateLimitException(String message, Double retryAfter, Integer limit, Integer remaining) {
        super(429, message, null);
        this.retryAfter = retryAfter;
        this.limit = limit;
        this.remaining = remaining;
    }

    public Double retryAfter() { return retryAfter; }
    public Integer limit() { return limit; }
    public Integer remaining() { return remaining; }
}
